src/servers/src/http.rs

```diff
@@ -1177,17 +1203,123 @@ mod test {
         let sql_instance = ServerSqlQueryHandlerAdapter::arc(instance.clone());
-        let server = HttpServerBuilder::new(HttpOptions::default())
+        let server = HttpServerBuilder::new(options)
             .with_sql_handler(sql_instance)
             .build();
-        server.build(server.make_app()).route(
+        server.build(server.make_app()).unwrap().route(
             "/test/timeout",
             get(forever.layer(ServiceBuilder::new().layer(timeout()))),
         )
     }
 
+    #[tokio::test]
+    pub async fn test_cors() {
+        // cors is on by default
+        let (tx, _rx) = mpsc::channel(100);
+        let app = make_test_app(tx);
+        let client = TestClient::new(app).await;
+
+        let res = client.get("/health").send().await;
+
+        assert_eq!(res.status(), StatusCode::OK);
+        assert_eq!(
+            res.headers()
+                .get(http::header::ACCESS_CONTROL_ALLOW_ORIGIN)
+                .expect("expect cors header origin"),
+            "*"
+        );
+
+        let res = client
+            .options("/health")
+            .header("Access-Control-Request-Headers", "x-greptime-auth")
+            .header("Access-Control-Request-Method", "DELETE")
+            .header("Origin", "https://example.com")
+            .send()
+            .await;
+        assert_eq!(res.status(), StatusCode::OK);
+        assert_eq!(
+            res.headers()
+                .get(http::header::ACCESS_CONTROL_ALLOW_ORIGIN)
+                .expect("expect cors header origin"),
+            "*"
+        );
+        assert_eq!(
+            res.headers()
+                .get(http::header::ACCESS_CONTROL_ALLOW_HEADERS)
+                .expect("expect cors header headers"),
+            "*"
+        );
+        assert_eq!(
+            res.headers()
+                .get(http::header::ACCESS_CONTROL_ALLOW_METHODS)
+                .expect("expect cors header methods"),
+            "GET,POST,PUT,DELETE,HEAD"
+        );
+    }
+
+    #[tokio::test]
+    pub async fn test_cors_custom_origins() {
+        // cors is on by default
+        let (tx, _rx) = mpsc::channel(100);
+        let origin = "https://example.com";
+
+        let options = HttpOptions {
+            cors_allowed_origins: vec![origin.to_string()],
+            ..Default::default()
+        };
+
+        let app = make_test_app_custom(tx, options);
+        let client = TestClient::new(app).await;
+
+        let res = client.get("/health").header("Origin", origin).send().await;
+
+        assert_eq!(res.status(), StatusCode::OK);
+        assert_eq!(
+            res.headers()
+                .get(http::header::ACCESS_CONTROL_ALLOW_ORIGIN)
+                .expect("expect cors header origin"),
+            origin
+        );
+
+        let res = client
+            .get("/health")
+            .header("Origin", "https://notallowed.com")
+            .send()
+            .await;
+
+        assert_eq!(res.status(), StatusCode::OK);
```

Comment:
A 403 Forbidden response would be more appropriate.

Reply:
This behaviour is controlled by tower's cors middleware. Also in cors, server doesn't reject requests, it's totally decided by client /browser.